Link to dataset: https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/02_Filtering_%26_Sorting/Euro12/Euro_2012_stats_TEAM.csv

In [1]:
# import libraries
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

In [2]:
# read csv file
df = pd.read_csv('https://raw.githubusercontent.com/guipsamora/pandas_exercises/master/02_Filtering_%26_Sorting/Euro12/Euro_2012_stats_TEAM.csv')

In [3]:
# to view all the columns
pd.set_option('display.max_columns', None)
df.head(10)

Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
0         Croatia      4               13                12             51.9%   
1  Czech Republic      4               13                18             41.9%   
2         Denmark      4               10                10             50.0%   
3         England      5               11                18             50.0%   
4          France      3               22                24             37.9%   
5         Germany     10               32                32             47.8%   
6          Greece      5                8                18             30.7%   
7           Italy      6               34                45             43.0%   
8     Netherlands      2               12                36             25.0%   
9          Poland      2               15                23             39.4%   

  % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
0            16.0%                          32             0              0   
1            12.9%                          39             0              0   
2            20.0%                          27             1              0   
3            17.2%                          40             0              0   
4             6.5%                          65             1              0   
5            15.6%                          80             2              1   
6            19.2%                          32             1              1   
7             7.5%                         110             2              0   
8             4.1%                          60             2              0   
9             5.2%                          48             0              0   

   Penalties not scored  Headed goals  Passes  Passes completed  \
0                     0             2    1076               828   
1                     0             0    1565              1223   
2                     0             3    1298              1082   
3                     0             3    1488              1200   
4                     0             0    2066              1803   
5                     0             2    2774              2427   
6                     1             0    1187               911   
7                     0             2    3016              2531   
8                     0             0    1556              1381   
9                     0             1    1059               852   

  Passing Accuracy  Touches  Crosses  Dribbles  Corners Taken  Tackles  \
0            76.9%     1706       60        42             14       49   
1            78.1%     2358       46        68             21       62   
2            83.3%     1873       43        32             16       40   
3            80.6%     2440       58        60             16       86   
4            87.2%     2909       55        76             28       71   
5            87.4%     3761      101        60             35       91   
6            76.7%     2016       52        53             10       65   
7            83.9%     4363       75        75             30       98   
8            88.7%     2163       50        49             22       34   
9            80.4%     1724       55        39             14       67   

   Clearances  Interceptions  Clearances off line  Clean Sheets  Blocks  \
0          83             56                  NaN             0      10   
1          98             37                  2.0             1      10   
2          61             59                  0.0             1      10   
3         106             72                  1.0             2      29   
4          76             58                  0.0             1       7   
5          73             69                  0.0             1      11   
6         123             87                  0.0             1      23   
7         137            136                  1.0             2      18   
8          41          

In [4]:
# to view shape of data
df.shape

(16, 35)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16 entries, 0 to 15
Data columns (total 35 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Team                        16 non-null     object 
 1   Goals                       16 non-null     int64  
 2   Shots on target             16 non-null     int64  
 3   Shots off target            16 non-null     int64  
 4   Shooting Accuracy           16 non-null     object 
 5   % Goals-to-shots            16 non-null     object 
 6   Total shots (inc. Blocked)  16 non-null     int64  
 7   Hit Woodwork                16 non-null     int64  
 8   Penalty goals               16 non-null     int64  
 9   Penalties not scored        16 non-null     int64  
 10  Headed goals                16 non-null     int64  
 11  Passes                      16 non-null     int64  
 12  Passes completed            16 non-null     int64  
 13  Passing Accuracy            16 non-nu

In [6]:
# to check for duplication
df.nunique()

Team                          16
Goals                          8
Shots on target               13
Shots off target              13
Shooting Accuracy             15
% Goals-to-shots              14
Total shots (inc. Blocked)    14
Hit Woodwork                   5
Penalty goals                  2
Penalties not scored           2
Headed goals                   4
Passes                        16
Passes completed              16
Passing Accuracy              15
Touches                       16
Crosses                       14
Dribbles                      15
Corners Taken                 13
Tackles                       14
Clearances                    16
Interceptions                 15
Clearances off line            3
Clean Sheets                   4
Blocks                        10
Goals conceded                 7
Saves made                    10
Saves-to-shots ratio          15
Fouls Won                     13
Fouls Conceded                14
Offsides                      12
Yellow Car

In [7]:
# to view null values
df.isnull().sum()

Team                          0
Goals                         0
Shots on target               0
Shots off target              0
Shooting Accuracy             0
% Goals-to-shots              0
Total shots (inc. Blocked)    0
Hit Woodwork                  0
Penalty goals                 0
Penalties not scored          0
Headed goals                  0
Passes                        0
Passes completed              0
Passing Accuracy              0
Touches                       0
Crosses                       0
Dribbles                      0
Corners Taken                 0
Tackles                       0
Clearances                    0
Interceptions                 0
Clearances off line           1
Clean Sheets                  0
Blocks                        0
Goals conceded                0
Saves made                    0
Saves-to-shots ratio          0
Fouls Won                     0
Fouls Conceded                0
Offsides                      0
Yellow Cards                  0
Red Card

In [8]:
# stastical description of data
df.describe().T

count         mean          std     min      25%  \
Goals                        16.0     4.750000     2.886751     1.0     2.75   
Shots on target              16.0    17.125000    10.582218     7.0     9.75   
Shots off target             16.0    24.937500    10.680005    10.0    18.00   
Total shots (inc. Blocked)   16.0    54.937500    26.065223    27.0    36.50   
Hit Woodwork                 16.0     1.250000     1.612452     0.0     0.00   
Penalty goals                16.0     0.187500     0.403113     0.0     0.00   
Penalties not scored         16.0     0.062500     0.250000     0.0     0.00   
Headed goals                 16.0     1.375000     1.024695     0.0     0.75   
Passes                       16.0  1763.375000   906.177898   851.0  1190.75   
Passes completed             16.0  1467.375000   827.580721   606.0   951.50   
Touches                      16.0  2579.187500  1122.164707  1433.0  1856.25   
Crosses                      16.0    57.187500    18.616189    33.0    43.75   
Dribbles                     16.0    52.312500    22.703065    18.0    37.25   
Corners Taken                16.0    21.562500    11.236659     7.0    14.00   
Tackles                      16.0    68.375000    22.823599    34.0    54.25   
Clearances                   16.0    86.375000    24.611312    41.0    73.75   
Interceptions                16.0    63.562500    25.765529    29.0    44.50   
Clearances off line          15.0     0.333333     0.617213     0.0     0.00   
Clean Sheets                 16.0     1.062500     1.289380     0.0     0.00   
Blocks                       16.0    12.562500     6.937519     4.0     8.00   
Goals conceded               16.0     4.750000     1.983263     1.0     3.00   
Saves made                   16.0    12.125000     4.573474     6.0     9.75   
Fouls Won                    16.0    52.937500    22.915697    25.0    35.75   
Fouls Conceded               16.0    55.625000    18.973227    30.0    44.50   
Offsides                     16.0     8.125000     4.910872     2.0     4.00   
Yellow Cards                 16.0     7.437500     3.265348     4.0     5.00   
Red Cards                    16.0     0.187500     0.403113     0.0     0.00   
Subs on                      16.0    10.875000     3.537890     7.0     8.50   
Subs off                     16.0    10.875000     3.537890     7.0     8.50   
Players Used                 16.0    17.250000     1.527525    15.0    16.00   

                               50%      75%     max  
Goals                          4.5     5.25    12.0  
Shots on target               13.0    22.00    42.0  
Shots off target              23.5    32.25    45.0  
Total shots (inc. Blocked)    44.0    68.75   110.0  
Hit Woodwork                   1.0     2.00     6.0  
Penalty goals                  0.0     0.00     1.0  
Penalties not scored           0.0     0.00     1.0  
Headed goals                   1.5     2.00     3.0  
Passes                      1522.0  1934.75  4317.0  
Passes completed            1211.5  1546.50  3820.0  
Touches                     2220.5  2921.25  5585.0  
Crosses                       53.5    62.25   101.0  
Dribbles                      51.0    65.00   106.0  
Corners Taken                 19.5    28.50    44.0  
Tackles                       65.0    80.00   122.0  
Clearances                    85.0    99.00   137.0  
Interceptions                 58.5    73.75   136.0  
Clearances off line            0.0     0.50     2.0  
Clean Sheets                   1.0     1.25     5.0  
Blocks                        10.0    13.50    29.0  
Goals conceded                 5.0     6.00     9.0  
Saves made                    11.0    13.50    22.0  
Fouls Won                     45.5    64.00   102.0  
Fouls Conceded                51.0    64.75    90.0  
Offsides                       7.5    11.25    19.0  
Yellow Cards                   6.5     9.00    16.0  
Red Cards                      0.0     0.00     1.0  
Subs on                       10.5 

In [9]:
# stastical description of object dtype
df.describe(include = 'O')

Team Shooting Accuracy % Goals-to-shots Passing Accuracy  \
count        16                16               16               16   
unique       16                15               14               15   
top     Croatia             50.0%            16.0%            83.9%   
freq          1                 2                2                2   

       Saves-to-shots ratio  
count                    16  
unique                   15  
top                   66.7%  
freq                      2

### 1. How many teams participated in the Euro2012?

In [10]:
df['Team'].nunique()

16

#### Total 16 teams participated.

### 2. What is the number of columns in the dataset?

In [11]:
len(df.columns)

35

### 3. View only the columns Team, Yellow Cards and Red Cards and assign them to a dataframe called discipline.

In [12]:
discipline = df[['Team', 'Yellow Cards', 'Red Cards']]
discipline

Team  Yellow Cards  Red Cards
0               Croatia             9          0
1        Czech Republic             7          0
2               Denmark             4          0
3               England             5          0
4                France             6          0
5               Germany             4          0
6                Greece             9          1
7                 Italy            16          0
8           Netherlands             5          0
9                Poland             7          1
10             Portugal            12          0
11  Republic of Ireland             6          1
12               Russia             6          0
13                Spain            11          0
14               Sweden             7          0
15              Ukraine             5          0

### 4. Sort the teams by Red Cards, then to Yellow Cards.

In [15]:
df.sort_values(by = ['Red Cards', 'Yellow Cards'])

Team  Goals  Shots on target  Shots off target  \
2               Denmark      4               10                10   
5               Germany     10               32                32   
3               England      5               11                18   
8           Netherlands      2               12                36   
15              Ukraine      2                7                26   
4                France      3               22                24   
12               Russia      5                9                31   
1        Czech Republic      4               13                18   
14               Sweden      5               17                19   
0               Croatia      4               13                12   
13                Spain     12               42                33   
10             Portugal      6               22                42   
7                 Italy      6               34                45   
11  Republic of Ireland      1                7                12   
9                Poland      2               15                23   
6                Greece      5                8                18   

   Shooting Accuracy % Goals-to-shots  Total shots (inc. Blocked)  \
2              50.0%            20.0%                          27   
5              47.8%            15.6%                          80   
3              50.0%            17.2%                          40   
8              25.0%             4.1%                          60   
15             21.2%             6.0%                          38   
4              37.9%             6.5%                          65   
12             22.5%            12.5%                          59   
1              41.9%            12.9%                          39   
14             47.2%            13.8%                          39   
0              51.9%            16.0%                          32   
13             55.9%            16.0%                         100   
10             34.3%             9.3%                          82   
7              43.0%             7.5%                         110   
11             36.8%             5.2%                          28   
9              39.4%             5.2%                          48   
6              30.7%            19.2%                          32   

    Hit Woodwork  Penalty goals  Penalties not scored  Headed goals  Passes  \
2              1              0                     0             3    1298   
5              2              1                     0             2    2774   
3              0              0                     0             3    1488   
8              2              0                     0             0    1556   
15             0              0                     0             2    1276   
4              1              0                     0             0    2066   
12             2              0                     0             1    1602   
1              0              0                     0             0    1565   
14             3              0                     0             1    1192   
0              0              0                     0             2    1076   
13             0              1                     0             2    4317   
10             6              0                     0             2    1891   
7              2              0                     0             2    3016   
11             0              0                     0             1     851   
9              0              0                     0             1    1059   
6              1              1                     1             0    1187   

    Passes completed Passing Accuracy  Touches  Crosses  Dribbles  \
2               1082            83.3%     1873       43        32   
5               2427            87.4%     3761      101        60   
3               1200            80.6%     2440       58        60   
8               1381            88.7%     2163     

### 5. Calculate the mean Yellow Cards given per Team.

In [16]:
df['Yellow Cards'].mean()

7.4375

#### The mean of yellow cards given per team is 7.44.

### 6. Filter teams that scored more than 6 goals.

In [22]:
df[df['Goals'] > 6]

Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
5   Germany     10               32                32             47.8%   
13    Spain     12               42                33             55.9%   

   % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
5             15.6%                          80             2              1   
13            16.0%                         100             0              1   

    Penalties not scored  Headed goals  Passes  Passes completed  \
5                      0             2    2774              2427   
13                     0             2    4317              3820   

   Passing Accuracy  Touches  Crosses  Dribbles  Corners Taken  Tackles  \
5             87.4%     3761      101        60             35       91   
13            88.4%     5585       69       106             44      122   

    Clearances  Interceptions  Clearances off line  Clean Sheets  Blocks  \
5           73             69                  0.0             1      11   
13         102             79                  0.0             5       8   

    Goals conceded  Saves made Saves-to-shots ratio  Fouls Won  \
5                6          10                62.6%         63   
13               1          15                93.8%        102   

    Fouls Conceded  Offsides  Yellow Cards  Red Cards  Subs on  Subs off  \
5               49        12             4          0       15        15   
13              83        19            11          0       17        17   

    Players Used  
5             17  
13            18

### 7. Select the teams that start with the letter G.

In [32]:
df[df['Team'].str.startswith('G')]

Team  Goals  Shots on target  Shots off target Shooting Accuracy  \
5  Germany     10               32                32             47.8%   
6   Greece      5                8                18             30.7%   

  % Goals-to-shots  Total shots (inc. Blocked)  Hit Woodwork  Penalty goals  \
5            15.6%                          80             2              1   
6            19.2%                          32             1              1   

   Penalties not scored  Headed goals  Passes  Passes completed  \
5                     0             2    2774              2427   
6                     1             0    1187               911   

  Passing Accuracy  Touches  Crosses  Dribbles  Corners Taken  Tackles  \
5            87.4%     3761      101        60             35       91   
6            76.7%     2016       52        53             10       65   

   Clearances  Interceptions  Clearances off line  Clean Sheets  Blocks  \
5          73             69                  0.0             1      11   
6         123             87                  0.0             1      23   

   Goals conceded  Saves made Saves-to-shots ratio  Fouls Won  Fouls Conceded  \
5               6          10                62.6%         63              49   
6               7          13                65.1%         67              48   

   Offsides  Yellow Cards  Red Cards  Subs on  Subs off  Players Used  
5        12             4          0       15        15            17  
6        12             9          1       12        12            20

### 8. Select the first 7 columns.

In [35]:
df.iloc[:,:7]

Team  Goals  Shots on target  Shots off target  \
0               Croatia      4               13                12   
1        Czech Republic      4               13                18   
2               Denmark      4               10                10   
3               England      5               11                18   
4                France      3               22                24   
5               Germany     10               32                32   
6                Greece      5                8                18   
7                 Italy      6               34                45   
8           Netherlands      2               12                36   
9                Poland      2               15                23   
10             Portugal      6               22                42   
11  Republic of Ireland      1                7                12   
12               Russia      5                9                31   
13                Spain     12               42                33   
14               Sweden      5               17                19   
15              Ukraine      2                7                26   

   Shooting Accuracy % Goals-to-shots  Total shots (inc. Blocked)  
0              51.9%            16.0%                          32  
1              41.9%            12.9%                          39  
2              50.0%            20.0%                          27  
3              50.0%            17.2%                          40  
4              37.9%             6.5%                          65  
5              47.8%            15.6%                          80  
6              30.7%            19.2%                          32  
7              43.0%             7.5%                         110  
8              25.0%             4.1%                          60  
9              39.4%             5.2%                          48  
10             34.3%             9.3%                          82  
11             36.8%             5.2%                          28  
12             22.5%            12.5%                          59  
13             55.9%            16.0%                         100  
14             47.2%            13.8%                          39  
15             21.2%             6.0%                          38

### 9. Select all columns except the last 3.

In [40]:
df.iloc[:, :-3]

Team  Goals  Shots on target  Shots off target  \
0               Croatia      4               13                12   
1        Czech Republic      4               13                18   
2               Denmark      4               10                10   
3               England      5               11                18   
4                France      3               22                24   
5               Germany     10               32                32   
6                Greece      5                8                18   
7                 Italy      6               34                45   
8           Netherlands      2               12                36   
9                Poland      2               15                23   
10             Portugal      6               22                42   
11  Republic of Ireland      1                7                12   
12               Russia      5                9                31   
13                Spain     12               42                33   
14               Sweden      5               17                19   
15              Ukraine      2                7                26   

   Shooting Accuracy % Goals-to-shots  Total shots (inc. Blocked)  \
0              51.9%            16.0%                          32   
1              41.9%            12.9%                          39   
2              50.0%            20.0%                          27   
3              50.0%            17.2%                          40   
4              37.9%             6.5%                          65   
5              47.8%            15.6%                          80   
6              30.7%            19.2%                          32   
7              43.0%             7.5%                         110   
8              25.0%             4.1%                          60   
9              39.4%             5.2%                          48   
10             34.3%             9.3%                          82   
11             36.8%             5.2%                          28   
12             22.5%            12.5%                          59   
13             55.9%            16.0%                         100   
14             47.2%            13.8%                          39   
15             21.2%             6.0%                          38   

    Hit Woodwork  Penalty goals  Penalties not scored  Headed goals  Passes  \
0              0              0                     0             2    1076   
1              0              0                     0             0    1565   
2              1              0                     0             3    1298   
3              0              0                     0             3    1488   
4              1              0                     0             0    2066   
5              2              1                     0             2    2774   
6              1              1                     1             0    1187   
7              2              0                     0             2    3016   
8              2              0                     0             0    1556   
9              0              0                     0             1    1059   
10             6              0                     0             2    1891   
11             0              0                     0             1     851   
12             2              0                     0             1    1602   
13             0              1                     0             2    4317   
14             3              0                     0             1    1192   
15             0              0                     0             2    1276   

    Passes completed Passing Accuracy  Touches  Crosses  Dribbles  \
0                828            76.9%     1706       60        42   
1               1223            78.1%     2358       46        68   
2               1082            83.3%     1873       43        32   
3               1200            80.6%     2440     

### 10. Present only the Shooting Accuracy from England, Italy and Russia.

In [45]:
countries = ['England', 'Italy', 'Russia']
filtered_df = df[df['Team'].isin(countries)]
shooting_accuracy = filtered_df[['Team', 'Shooting Accuracy']]
shooting_accuracy

Team Shooting Accuracy
3   England             50.0%
7     Italy             43.0%
12   Russia             22.5%